### PR TITLE
hw-33

### PR DESCRIPTION
**Pull request by Rommelua**

### CODE REVIEW

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        }
+    }
+
+    @Override
+    public Order update(Order order) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            deleteProductsFromDB(order.getId(), connection);
+            addProductsToDB(order, connection);
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE orders SET user_id = ? "
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:97]: do not think we can change user for order. Is it possible?

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    private Order getOrderFromResultSet(ResultSet resultSet) throws SQLException {
+        long id = resultSet.getLong("id");
+        long userId = resultSet.getLong("user_id");
+        List<Product> products = new ArrayList<>();
+        do {
+            products.add(new Product(resultSet.getLong("product_id"),
+                    resultSet.getString("name"), resultSet.getDouble("price")));
+        } while (resultSet.next() && id == resultSet.getLong("id"));
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:128]: what is the purpose of do while here? is it possible that we have an empty result set?

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        }
+    }
+
+    @Override
+    public ShoppingCart update(ShoppingCart shoppingCart) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            deleteProductsFromDB(shoppingCart.getId(), connection);
+            addProductsToDB(shoppingCart, connection);
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE shopping_carts SET user_id = ? "
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java:100]: can you update the user in the shopping cart if it should be created automatically? I think it makes no sense

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        }
+        return Optional.empty();
+    }
+
+    private ShoppingCart getShoppingCartFromResultSet(ResultSet resultSet) throws SQLException {
+        long id = resultSet.getLong("id");
+        long userId = resultSet.getLong("user_id");
+        List<Product> products = new ArrayList<>();
+        long productId = resultSet.getLong("product_id");
+        if (productId != 0) {
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java:140]: what do you mean by this check? AUTO_INCREMENT in SQL makes count from 1, so this statement will always be true

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                user.setId(resultSet.getLong("id"));
+            }
+            addRolesToDB(user, connection);
+            return user;
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:41]: Connecting to what you have written to me. In this case user has roles, but id of these roles = null. I am sure we do not need to use id at the point, but it is not okay to have nulls in our project :)

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                    "SELECT * FROM users u "
+                    + "JOIN users_roles ur ON u.id = ur.user_id "
+                    + "JOIN roles r on ur.role_id = r.id "
+                    + "WHERE deleted = FALSE ORDER BY u.id",
+                    ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_UPDATABLE);
+            ResultSet resultSet = statement.executeQuery();
+            List<User> allUsers = new ArrayList<>();
+            while (resultSet.next()) {
+                allUsers.add(getUserFromResultSet(resultSet));
+                resultSet.previous();
```
Comment (andromash) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:98]: so you stay at the same point all the time?

**Review by andromash — COMMENTED** [review]
Good job!

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 1 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:97]: In the real shop it's really difficult to imagine thise situation. But I tried to update all the fields. If mentor doesn't mind I will remove extra logic.

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 2 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:128]: Can't understand your comment. Of course I don't expect empty set. While is necessary to call next() multiple times.

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 3 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java:100]: same as order

File: src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java (hunk 4 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java:140]: when cart is empty this field is null and getLong() returns 0

File: src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java (hunk 1 above)
Comment (RomanDubovskyi) [reply src/main/java/com/internet/shop/dao/jdbc/OrderDaoJdbcImpl.java:97]: it's ok the way it is right now, no need to make logic overcomplicated for this project

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 5 above)
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:41]: does this user have roles_id initialized?

Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:41]: _"but id of these roles = null"_ why do you think so???

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 6 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:98]: I am doing step back after the last calling next() at line 159

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                    Statement.RETURN_GENERATED_KEYS);
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                user.setId(resultSet.getLong("id"));
+            }
+            addRolesToDB(user, connection);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:37]: it's ok to pass your connection into other method, but let's close unneded statement, cause we create a new one in addRolesToDB() method
same in other cases

Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:37]: fixed

File: src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java (hunk 5 above)
Comment (Rommelua) [reply src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java:41]: fixed

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**